chapter4.adoc

```diff
@@ -221,28 +221,70 @@ follows:
-for the resulting physical address may be checked (and possibly cached) at any
-point between the address translation and the explicit memory access. If
-caching is occuring, when the `MTT` settings are modified, `M-mode` software
-must synchronize the cached `MTT` state with the virtual memory system and any
-`PMP`, `MTT` or address-translation caches. This is accomplished by executing
-an `SFENCE.VMA` instruction with `rs1=x0` and `rs2=x0`, or `HFENCE.GVMA` as
-needed, after the `MTT` is modified. If page-based virtual memory is not
-implemented, memory accesses check the `PMP` settings synchronously, but may
-check `MTT` settings that are cached, so a `MTT` invalidation (`MTTINVAL`)
-instruction is needed. When Svinval is implemented, `MTTINVAL` is only ordered
-against `SFENCE.W.INVAL` and `SFENCE.INVAL.IR` instructions. As part of the
-`MTT` update, the RDSM must ensure that it uses `SFENCE.W.INVAL` to guarantee
-that any previous stores to `MTT` are made visible before invoking the
-`MTTINVAL`. The RDSM must then use `SFENCE.INVAL.IR` to guarantee that all
-subsequent implicit references to `MTT` are ordered to be after the `MTT` cache
-invalidation.
-
-_[TBD - register interface for flushing all MTT cached entries, vs specific
-physical address at page size granularity]._
-
+=== Access Enforcement and Fault Reporting
+
+As shown in <<mtt-lookup>>, MTT lookup composes with, but does not require,
+page-based virtual memory (MMU, IOMMU) and physical memory protection mechanisms
+(PMP, Smepmp, IOPMP). When paging is enabled, instructions that access virtual
+memory may result in multiple physical-memory accesses, including (implicit
+S-mode) accesses to the page tables. MTT checks also apply to these implicit
+S-mode accesses - those accesses will be treated as reads for translation and as
+writes when A/D bits are updated in page table entries when `Svadu` is
+implemented.
+
+MTT is checked for all accesses to physical memory, unless the effective privilege
+mode is M, including accesses that have undergone virtual to physical memory
+translation, but excluding MTT structure accesses. Data accesses in M-mode
+when the MPRV bit in mstatus is set and the MPP field in mstatus contains S
+or U are subject to MTT checks. MTT structure accesses are subject to PMP/
+Smepmp and IOPMP checks. The MTT checker indexes the MTT using the
+physical address of the access to lookup and enforce the access permissions.
+A mismatch of the access type and the access permissions specified in the
+MTT entry that applies to the accessed region is reported as a trap to the
+RDSM which may report it to a supervisor domain. To enable composing
+with Sv modes, the MTT supports configuration at supported architectural
+page sizes. MTT violations manifest as instruction, load, or store access-fault
+exceptions. The exception conditions for MTT are checked when the access
+to memory is performed.
+
+=== Caching of MTT and Supervisor Domain Fence Instruction
```

Comment:
This section is mostly redundant and covered by the MFENCE.SPA and MINVAL.SPA. Whether MTT is cached is microarchitectural.

Reply:
i thought of keeping this section since it is in the normative Smmtt part of the spec. if anything is incorrect here we should fix it, but o/w it seems useful.